glean-core/ffi/src/lib.rs

```diff
@@ -37,17 +37,28 @@ lazy_static! {
 type RawStringArray = *const *const c_char;
 
 /// Create a vector of strings from a raw C-like string array
-unsafe fn from_raw_string_array(arr: RawStringArray, len: i32) -> Vec<String> {
-    if arr.is_null() || len == 0 {
-        return vec![];
+///
+/// ## Safety
+///
+/// * We check the array pointer for validity (non-null)
+/// * FfiStr checks each individual char pointer for validity (non-null)
+/// * We discard invalid char pointers (null pointer)
+/// * invalid UTF-8 is replaced with the replacement character (guaranteed by ffi-support)
+fn from_raw_string_array(arr: RawStringArray, len: i32) -> Vec<String> {
+    unsafe {
+        if arr.is_null() || len == 0 {
+            return vec![];
+        }
+
+        let arr_ptrs = std::slice::from_raw_parts(arr, len as usize);
+        arr_ptrs
+            .iter()
+            .flat_map(|&p| {
+                // Drop invalid strings
+                FfiStr::from_raw(p).into_opt_string()
```

Comment:
I assume the typical failure here would be invalid UTF-8 codepoints?  I'm not sure whether the best behavior is to 

1) drop invalid strings (as here)
2) replace them with empty strings
3) fail the entire conversion (by returning an error result)

I could see this introducing a whole other class of error downstream if the array of strings suddenly had a different length than expected.  I think I would lean on option 3 here, with 2 being a close second.  Though a lot of this depends on expected usage -- but it's awfully hard for the caller to make sense of this and act accordingly from the outside here.

Reply:
Right.
Up to now glean-core just ... explodes on invalid UTF-8 codepoints, so we're not much better off.

And acutally, we're not dropping strings with invalid UTF-8 codepoints, we replace those codepoints:

>  Note that unlike FfiStr::as_opt_str, invalid UTF-8 is replaced with the replacement character instead of causing us to return None.

So ... I guess we should decide first what we want to do. I'm very much ok with returning a result (and thus bailing out early)